osquery/tables/ipmi/ipmi_info.cpp

```diff
@@ -0,0 +1,1309 @@
+ * table.  The initialization process waits for OpenIPMI to be in the "fully up"
+ * state so it can a few minutes.
+ */
+class IPMIClient : public InternalRunnable {
+ public:
+  /**
+   * @brief retrieves instance of IPMIClient (singleton).
+   *
+   */
+  static IPMIClient& get();
+
+ public:
+  /**
+   * @brief retrieves QueryData for ipmi_threshold_sensors.
+   *
+   * @param results Reference to QueryData.
+   */
+  void getThresholdSensors(QueryData& results);
+
+  /**
+   * @brief retrieves QueryData for ipmi_lan.
+   *
+   * @param results Reference to QueryData.
+   */
+  void getLANConfigs(QueryData& results);
+
+  /**
+   * @brief retrieves QueryData for ipmi_fru.
+   *
+   * @param results Reference to QueryData.
+   */
+  void getFRUs(QueryData& results);
+
+  /**
+   * @brief retrieves QueryData for ipmi_mc.
+   *
+   * @param results Reference to QueryData.
+   */
+  void getMCs(QueryData& results);
+
+  /**
+   * @brief checks if IPMIClient is successfully initiated.
+   *
+   * @return bool indicating client state.
+   */
+  bool up();
+
+  /// Starts background clean up routine.  Implements InternalRunnable.
+  void start() override;
+
+  /// Sets running state to false and does one final clean up.
+  void stop() override;
+
+  ~IPMIClient();
+  IPMIClient(IPMIClient const& client) = delete;
+  void operator=(IPMIClient const& client) = delete;
+
+ private:
+  IPMIClient();
+
+  /// Sets domain of client instance.
+  void setDomain(ipmi_domain_t* d);
+
+  /// Sets the local IPMI LAN channel.
+  void setLANCh(unsigned int ch);
+
+  /// Gets a ipmi_lanparm_t* by key name from the instance.
+  ipmi_lanparm_t* getLANParm(const std::string& name, ipmi_mc_t* mc);
+
+  /// Add a ipmi_lanparm_t* by key name to the instance.
+  bool addLANParm(const std::string& name, ipmi_mc_t* mc);
+
+  /// Insert a column pair to the Row at key.
+  void insertRowsQueue(std::string key,
+                       const std::string& columnName,
+                       const std::string& columnValue);
+
+  /// Insert a new Row for the key.
+  void insertRowsQueue(const std::string& key, const Row& row);
+
+  /// Push a new parmData* to instance property parmsOnHeap_.
+  void pushParmData(parmData* parm);
+
+  /// Removes parmData* from parmsOnHeap_ and calls delete on it.
+  void rmParmData(parmData* parm);
+
+  /// Performs one operation to get an IPMI event.  Times out after 500ms.
+  int oneOp();
+
+  /// Blocks for duration specified by timeoutDurMS or termination function
+  /// returns true performing oneOp.
+  void blkAndOp(std::function<bool()> termF, int timeoutDurMS = 500);
+
+  /// Searches all IPMI channels for 802.3 LAN channel; uses first found.
+  void findLANCh();
+
+  /// Converts instance local rowsQueue_ to QueryData.
+  void toQueryData(QueryData& results);
+
+  /// Iterate all IPMI entities by registering cb.
+  void iterateEntities(ipmi_entities_iterate_entity_cb cb);
```

Comment:
so `ipmi_entities_iterate_entity_cb` is actually `typedef void (*ipmi_entities_iterate_entity_cb)(ipmi_entity_t *entity, void *cb_data);` .  AFAIK, there aren't any benefits to passing a function pointer by reference (I guess unless I wanted to modify what `cb` points to)..Thoughts?